File: manuals/lua/mergesort.lua
function mergeSort(A, p, r)
    if p < r then
        local q = math.floor((p + r) / 2)
        mergeSort(A, p, q)
        mergeSort(A, q + 1, r)
        merge(A, p, q, r)
    end
end

function merge(A, p, q, r)
    local n1 = q - p + 1
    local n2 = r - q
    local left = {}
    local right = {}

    for i = 1, n1 do
        left[i] = A[p + i - 1]
    end
    for i = 1, n2 do
        right[i] = A[q + i]
    end

    left[n1 + 1] = math.huge
    right[n2 + 1] = math.huge

    local i = 1
    local j = 1

    for k = p, r do
        if left[i] <= right[j] then
            A[k] = left[i]
            i = i + 1
        else
            A[k] = right[j]
            j = j + 1
        end
    end
end

function printArray(A)
    for i = 1, #A do
        io.write(A[i] .. " ")
    end
    io.write("\n")
end

A = {11, 45, 14, 19, 19, 8, 10}

io.write("Original:\n")
printArray(A)
mergeSort(A, 1, #A)
io.write("Sorted:\n")
printArray(A)

File: manuals/java/mergesort.java
public class MergeSortTest {

  public static void main(String[] args) {
    int A[] = { 11, 45, 14, 19, 19, 8, 10 };
    System.out.println("Original:");
    for (int a : A) {
      System.out.print(a + " ");
    }
    System.out.println();
    mergeSort(A);
    System.out.println("Sorted:");
    for (int a : A) {
      System.out.print(a + " ");
    }
    System.out.println();
  }

  public static void mergeSort(int[] A) {
    sort(A, 0, A.length - 1);
  }

  public static void sort(int[] A, int start, int end) {
    if (start >= end) return;
    int mid = (start + end) / 2;
    sort(A, start, mid);
    sort(A, mid + 1, end);
    merge(A, start, mid, end);
  }

  public static void merge(int[] A, int start, int mid, int end) {
    int[] temp = new int[A.length];
    int k = 0;
    int i = start;
    int j = mid + 1;
    while (i <= mid && j <= end) {
      if (A[i] <= A[j]) {
        temp[k++] = A[i++];
      } else {
        temp[k++] = A[j++];
      }
    }
    while (i <= mid) {
      temp[k++] = A[i++];
    }
    while (j <= end) {
      temp[k++] = A[j++];
    }
    for (int m = 0; m < k; m++) {
      A[m + start] = temp[m];
    }
  }
}
